### PR TITLE
Make doc CI use nightly

### LINKED ISSUE #15427: `cargo run -p ci` doesn't handle doc tests properly
> Maybe the command that you used didn't correctly export the env vars?

Ah, this command did the trick:
`$env:RUSTFLAGS='--cfg docsrs_dep'; $env:RUSTDOCFLAGS='--cfg=docsrs'; cargo +nightly doc`

Fixing the errors now!

_Originally posted by @BenjaminBrienen in https://github.com/bevyengine/bevy/issues/15387#issuecomment-2374009554_

### PR DESCRIPTION
**Pull request by BenjaminBrienen**
# Objective

Fixes #15427
Follow-up to #15428

## Solution

Use nightly and add the environment variables as suggested here: https://github.com/bevyengine/bevy/pull/15428#pullrequestreview-2331294421

## Testing

CI should run

### CODE REVIEW

**Comment by BenjaminBrienen** [comment]
@alice-i-cecile I think fixing that error would require making the whole CI use nightly. Do you have a different idea?

**Review by alice-i-cecile — APPROVED** [review]

**Comment by BD103** [comment]
Because docs are already built with nightly, I'm fine with enabling it here:

https://github.com/BenjaminBrienen/bevy/blob/2a9c0fad9a90c74a19e3820581b6e46de1fe8769/.github/workflows/ci.yml#L319

```diff
- - uses: dtolnay/rust-toolchain@stable
+ - uses: dtolnay/rust-toolchain@master
+   with:
+     toolchain: ${{ env.NIGHTLY_TOOLCHAIN }}
```

**Comment by mockersf** [comment]
the tool shouldn't force CI but use the version decided by the user

**Comment by BD103** [comment]
> The tool shouldn't force nightly but use the version decided by the user
> 
> They can print a warning that nightly is recommended

I agree. I think the path forward here is:

1. Make the CI checks for docs specifically use nightly.
2. Detect the version of `rustc` (run `rustc --version`) and enable nightly features based on that.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -127,7 +127,7 @@ jobs:
       - name: Check Compile
         # See tools/ci/src/main.rs for the commands this runs
         run: cargo run -p ci -- compile
-
```
Comment (BenjaminBrienen) [inline .github/workflows/ci.yml:130]: We should look into CI enforced yml formatting

**Review by BD103 — APPROVED** [review]
The changes look good, but is it possible to test them before this is merged? (Perhaps test it on your fork?)

**Comment by BenjaminBrienen** [comment]
> The changes look good, but is it possible to test them before this is merged? (Perhaps test it on your fork?)

Can you explain what you mean by that? I'm not sure what I could do on my fork that would be different from the CI here.

**Comment by BD103** [comment]
> Can you explain what you mean by that? I'm not sure what I could do on my fork that would be different from the CI here.

Ah my bad, I thought this PR changed [`docs.yml`](https://github.com/bevyengine/bevy/blob/main/.github/workflows/docs.yml), which wouldn't be caught by CI. Sorry about that, @alice-i-cecile you're free to merge!

**Pull request merged**